ravenpy/utilities/data_assimilation.py

```diff
@@ -191,3 +194,173 @@ def update_state(x, qobs_pert, qobs_error, qsim):
     xa = np.maximum(xa, 0)
 
     return xa
+
+
+def perturb_full_series(model, std, start_date, end_date, dists, n_members=25):
+    """
+    This function will take a timeseries as input and perturb each variable
+    according to the desired uncertainty/confidence level and distribution of
+    each variable. Returns an Xarray Dataset object with n_members x time values
+    for each variable.
+
+    Parameters
+    ----------
+    model : Raven model class object
+        The model that will be used to perform the simulations and assimilation
+    std : tuple
+        Pairs of variable: standard deviation of the uncertainty.
+    start_date : datetime object
+        Start date of the assimilation run.
+    end_date : datetime object
+        End date of the assimilation run.
+    dists : tuple
+        Pairs of variable:distribution to identify the type of distribution each variable follows.
+    n_members : integer
+        Number of members to use in the Ensemble Kalman Filter.
+
+    Returns
+    -------
+    perturbed: path object (string)
+        Path to the perturbed hydrometeorological timeseries for the entire period.
+    """
+
+    # Use the same random seed for both tasmin and tasmax
+    rs = np.random.SeedSequence(None).generate_state(1)[0]
+    seed = {"tasmin": rs, "tasmax": rs}
+
+    # === Create perturbed time series for full assimilation period ====
+    perturbed = {}
+    for key, s in std.items():
+        nc = model.config.rvt._var_cmds[key]
+
+        with xr.open_dataset(nc.file_name_nc) as ds:
+            da = ds.get(nc.var_name_nc).sel(time=slice(start_date, end_date))
+
+            perturbed[key] = perturbation(
+                da,
+                dists.get(key, "norm"),
+                std=s,
+                seed=seed.get(key, None),
+                member=n_members,
+            )
+
+            # Save flow for later
+            if key == "water_volume_transport_in_river_channel":
+                q_obs = da
```

Comment:
This is a bit weird.

Reply:
I agree, and I am not too sure how to handle this. At one point we "force" users to use this key to make sure there is Qobs in the netcdf. But I see here that the variable name we are using is "qobs" in our netcdf file, so I am guessing this is the long_name? But I can't seem to access the array by long_name, just the variable name. So I am a bit stumped on how to make this better. Ideas?

Reply:
No, what I'm confused about is that we return the perturbed dataset, including flow, + the original flow. I think the code would be cleaner if we just returned perturbed.